File: leet_code/linked_lists/rotate_list.js
const rotateList = (head, k) => {
  if (!head) return null;
  if (!head.next) return head;

  let tail = head,
      length = 1,
      step = 1;

  while (tail.next) {
    length++;
    tail = tail.next;
  }

  if ((length - k) % length === 0) return head;

  let cur = head,
      newHead = head.next;

  while (step < ((length - k) % length + length) % length ) {
    cur = cur.next;
    newHead = newHead.next;
    step++;
  }
  cur.next = null;
  tail.next = head;
  return newHead;
};

File: leet_code/linked_lists/rotate_list.rb
def rotate_list(head, k)
    return nil if !head
    return head if !head.next

    tail = head
    length, step = 1, 1

    while tail.next
        length += 1
        tail = tail.next
    end

    return head if ((length - k) % length) == 0

    cur, new_head = head, head.next

    while step < ((length - k) % length)
        cur, new_head = cur.next, new_head.next
        step += 1
    end

    cur.next, tail.next = nil, head
    new_head
end
